### PR TITLE
openjpeg: add meson build

### PR DESCRIPTION
**Pull request by dabrain34**
Build support:
- thirdparties(z, tiff, png,cms2)
- jp2
- jpip
- jpwl
- mj2
- jp3d

### CODE REVIEW

**Comment by dabrain34** [comment]
Thx for your comments on the issue.

I changed the dep's name for every libs.

This build targets the 2.3.1 version of openjpeg

**Comment by dabrain34** [comment]
@legeana ^

File: thirdparty/libtiff/meson.build [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+  if not c_inline
+    result = cc.compiles(code, name : 'basic inline', args : c_args)
+    if result
+      c_tif_config_data.set('INLINE_KEYWORD', keyword)
+      c_inline = true
+    endif
+  endif
+endforeach
+
+configure_file(input: 'tif_config.h.cmake.in', output : 'tif_config.h', configuration : c_tif_config_data, format : 'cmake@')
```
Comment (legeana) [inline thirdparty/libtiff/meson.build:306]: The build broke here with some very obscure error:

```
$ mesonwrap review libopenjp2 --pull-request=1 --approve
Inspecting project libopenjp2, branch 2.3.1.
Repo name valid: ☑
Branch name valid: ☑
Target branch is not master: ☑
Has readme.txt: ☑
Has LICENSE.build: ☑
Has upstream.wrap: ☑
Repo contains only buildsystem files: ☑
upstream.wrap has directory: ☑
upstream.wrap directory has no path separators: ☑
upstream.wrap has source_url: ☑
upstream.wrap has source_filename: ☑
upstream.wrap source_filename has no path separators: ☑
upstream.wrap has source_hash: ☑
upstream.wrap has source_url with version substring: ☑
Download URL works: ☑
Hash matches: ☑
upstream.wrap directory 'openjpeg-2.3.1' exists: ☑
Patch merges with source: ☑
The Meson build system
Version: 0.54.0
Source dir: /tmp/tmp4yri2fc_/src/openjpeg-2.3.1
Build dir: /tmp/tmp4yri2fc_/bin
Build type: native build
Project name: openjp2
Project version: 2.3.1
C compiler for the host machine: cc (gcc 9.3.0 "cc (Arch Linux 9.3.0-1) 9.3.0")
C linker for the host machine: cc ld.bfd 2.34
C++ compiler for the host machine: c++ (gcc 9.3.0 "c++ (Arch Linux 9.3.0-1) 9.3.0")
C++ linker for the host machine: c++ ld.bfd 2.34
Host machine cpu family: x86_64
Host machine cpu: x86_64
Library m found: YES
Run-time dependency threads found: YES
Checking for function "clock_gettime" : YES 
wx-config found: YES (/usr/bin/wx-config) 3.0.4
Run-time dependency WxWidgets found: YES 3.0.4
Found pkg-config: /usr/bin/pkg-config (1.6.3)
Run-time dependency zlib found: YES 1.2.11
Run-time dependency libpng found: YES 1.6.37
Message: STATUS: Your system seems to have a png lib available, we will use it to generate PNG lib
Found CMake: /usr/bin/cmake (3.17.0)
Run-time dependency libtiff found: NO (tried pkgconfig and cmake)
Run-time dependency libjpeg found: YES 2.0.4
Has header "assert.h" : YES 
Has header "ctype.h" : YES 
Has header "dlfcn.h" : YES 
Has header "fcntl.h" : YES 
Has header "float.h" : YES 
Has header "inttypes.h" : YES 
Has header "ieeefp.h" : NO 
Has header "io.h" : NO 
Has header "limits.h" : YES 
Has header "malloc.h" : YES 
Has header "math.h" : YES 
Has header "memory.h" : YES 
Has header "search.h" : YES 
Has header "stdarg.h" : YES 
Has header "stdint.h" : YES 
Has header "stdlib.h" : YES 
Has header "stdio.h" : YES 
Has header "string.h" : YES 
Has header "strings.h" : YES 
Has header "sys/stat.h" : YES 
Has header "sys/time.h" : YES 
Has header "sys/types.h" : YES 
Has header "time.h" : YES 
Has header "unistd.h" : YES 
Has header "windows.h" : NO 
Has header "winsock2.h" : NO 
Checking for function "getopt" : YES 
Checking for function "isascii" : YES 
Checking for function "memmove" : YES 
Checking for function "memset" : YES 
Checking for function "mmap" : YES 
Checking for function "strcasecmp" : YES 
Checking for function "strchr" : YES 
Checking for function "strrchr" : YES 
Checking for function "strstr" : YES 
Checking for function "strtol" : YES 
Checking for function "strtoul" : YES 
Checking for function "strtoull" : YES 
Checking for function "snprintf" : YES 
Checking for size of "signed int" : 4
Checking for size of "signed long" : 8
Checking for size of "signed long long" : 8
Checking for size of "signed short" : 2
Checking for size of "unsigned int" : 4
Checking for size of "unsigned long" : 8
Checking for size of "unsigned long long" : 8
Checking for size of "unsigned short" : 2
Checking for size of "unsigned char*" : 8
Checking for size of "ssize_t" : 8
Checking for function "floor" : YES 
Checking for function "sqrt" : YES 
Checking for function "pow" : YES 
Checking for function "lfind" : YES 
Checking for function "setmod" : NO 
Checking if "basic inline" compiles: YES 
Configuring tif_config.h using configuration
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/mesonbuild/mesonmain.py", line 131, in run
    return options.run_func(options)
  File "/usr/lib/python3.8/site-packages/mesonbuild/msetup.py", line 245, in run
    app.generate()
  File "/usr/lib/python3.8/site-packages/mesonbuild/msetup.py", line 159, in generate
    self._generate(env)
  File "/usr/lib/python3.8/site-packages/mesonbuild/msetup.py", line 192, in _generate
    intr.run()
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreter.py", line 4359, in run
    super().run()
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 441, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 466, in evaluate_codeblock
    raise e
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 459, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 472, in evaluate_statement
    return self.function_call(cur)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 829, in function_call
    return func(node, func_args, self.kwargs_string_keys(kwargs))
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 326, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 212, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreter.py", line 3881, in func_subdir
    self.evaluate_codeblock(codeblock)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 466, in evaluate_codeblock
    raise e
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 459, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 482, in evaluate_statement
    return self.evaluate_if(cur)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 566, in evaluate_if
    self.evaluate_codeblock(node.elseblock)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 466, in evaluate_codeblock
    raise e
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 459, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 482, in evaluate_statement
    return self.evaluate_if(cur)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 566, in evaluate_if
    self.evaluate_codeblock(node.elseblock)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 466, in evaluate_codeblock
    raise e
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 459, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 472, in evaluate_statement
    return self.function_call(cur)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 829, in function_call
    return func(node, func_args, self.kwargs_string_keys(kwargs))
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 326, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 212, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreter.py", line 3881, in func_subdir
    self.evaluate_codeblock(codeblock)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 466, in evaluate_codeblock
    raise e
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 459, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 472, in evaluate_statement
    return self.function_call(cur)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 829, in function_call
    return func(node, func_args, self.kwargs_string_keys(kwargs))
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 326, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 326, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 326, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  [Previous line repeated 2 more times]
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreterbase.py", line 212, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3.8/site-packages/mesonbuild/interpreter.py", line 4079, in func_configure_file
    mesonlib.do_conf_file(inputs_abs[0], ofile_abs, conf.held_object,
  File "/usr/lib/python3.8/site-packages/mesonbuild/mesonlib.py", line 964, in do_conf_file
    line, missing = do_replacement(regex, line, variable_format, confdata)
  File "/usr/lib/python3.8/site-packages/mesonbuild/mesonlib.py", line 880, in do_replacement
    assert variable_format == 'meson'
AssertionError
meson setup: ☒
```

Comment (legeana) [reply thirdparty/libtiff/meson.build:306]: I filed https://github.com/mesonbuild/meson/issues/6974. I will look into it later.

**Comment by dabrain34** [comment]
yes this is a known issue i fixed in meson https://github.com/mesonbuild/meson/pull/6917

Should work with meson master branch

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,150 @@
+
+project('openjp2', 'c', 'cpp', version : '2.3.1',
+        meson_version : '>= 0.46.0',
```
Comment (legeana) [inline meson.build]: Please set the minimum version to `0.55.0` or the lowest that includes https://github.com/mesonbuild/meson/pull/6917.

**Comment by legeana** [comment]
I am not sure what is the correct way to resolve this.

```
../../tmp/tmpxx9t1y1c/src/openjpeg-2.3.1/meson.build:2:0: ERROR: Meson version is 0.54.999 but project requires >= 0.55.0
```
I tried checking this with the HEAD meson. Maybe `0.54.999` version is the right one to require.

**Comment by dabrain34** [comment]
I changed it to 0.54.99. We can update after the next release.

Do you know when ?

**Pull request merged**